### PR TITLE
Feature/finite difference

### PR DESCRIPTION
**Pull request by SylvainBertrand**
This PR adds simple 1st order finite difference and integration for the different primitives in euclid.
The new stuff is available from `EuclidCoreTools`.
(I'm thinking about slowly migrating all the tools there so we can reduce the number of tool classes.)

Lemme know whatcha think.

### CODE REVIEW

File: src/main/java/us/ihmc/euclid/tools/EuclidCoreTools.java [hunk 1]
```diff
@@ -1403,4 +1421,351 @@ public static int previous(int index, int listSize)
    {
       return wrap(index - 1, listSize);
    }
+
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/euclid/tools/EuclidCoreTools.java:1424]: 👇 new methods for doing finite difference  👇

File: src/main/java/us/ihmc/euclid/tools/EuclidCoreTools.java [hunk 2]
```diff
@@ -1403,4 +1421,351 @@ public static int previous(int index, int listSize)
...
+                                       RigidBodyTransformReadOnly currentTransform,
+                                       double dt,
+                                       Vector3DBasics angularVelocityToPack,
+                                       Vector3DBasics linearVelocityToPack)
+   {
+      finiteDifference(previousTransform.getRotation(), currentTransform.getRotation(), dt, angularVelocityToPack);
+      finiteDifference(previousTransform.getTranslation(), currentTransform.getTranslation(), dt, linearVelocityToPack);
+   }
+
+   /**
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/euclid/tools/EuclidCoreTools.java:1613]: 👇 new methods for integrating 👇

File: src/main/java/us/ihmc/euclid/tools/EuclidCoreRandomTools.java [hunk 3]
```diff
@@ -1556,11 +1555,9 @@ public static void randomizeAxisAngle(Random random, AxisAngleBasics axisAngleTo
     */
    public static void randomizeAxisAngle(Random random, double minMaxAngle, AxisAngleBasics axisAngleToRandomize)
    {
-      // Generate uniformly random point on unit sphere (based on http://mathworld.wolfram.com/SpherePointPicking.html )
-      double height = 2.0 * random.nextDouble() - 1.0;
+      // Generate uniformly random point on the unit-sphere (based on http://mathworld.wolfram.com/SpherePointPicking.html )
       double angle = nextDouble(random, minMaxAngle);
-      double radius = EuclidCoreTools.squareRoot(1.0 - height * height);
-      axisAngleToRandomize.set(radius * EuclidCoreTools.cos(angle), radius * EuclidCoreTools.sin(angle), height, angle);
+      axisAngleToRandomize.set(EuclidCoreRandomTools.nextVector3D(random), angle);
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/euclid/tools/EuclidCoreRandomTools.java:1560]: The old implementation was incorrect in the sense that it wouldn't generate random axes that would be uniformly distributed.
Fixing this seems to break other tests. I'll work on fixing them

**Review by calvertdw — APPROVED** [review]
Holy moly that's a lot of stuff. Looks great!

File: src/test/java/us/ihmc/euclid/axisAngle/AxisAngleTest.java [hunk 4]
```diff
@@ -237,7 +233,7 @@ public AxisAngle createRandomAxisAngle(Random random)
    @Override
    public double getEpsilon()
    {
-      return 1.0e-12;
+      return 1.1e-12;
```
Comment (james-p-foster) [inline src/test/java/us/ihmc/euclid/axisAngle/AxisAngleTest.java:236]: Why this change? Purely curious.

File: src/test/java/us/ihmc/euclid/matrix/RotationMatrixTest.java [hunk 5]
```diff
@@ -1106,7 +1101,7 @@ public void testNumericalError() throws Exception
          RotationMatrix unnormalized = new RotationMatrix();
          RotationMatrix normalized = new RotationMatrix();
 
-         for (int i = 0; i < 268000; i++)
+         for (int i = 0; i < 90000; i++)
```
Comment (james-p-foster) [inline src/test/java/us/ihmc/euclid/matrix/RotationMatrixTest.java:1104]: Why this number of iterations?

**Review by james-p-foster — APPROVED** [review]
Looks good to me! Just curious about a few constants, but go for merge! 👍🏻

File: src/test/java/us/ihmc/euclid/axisAngle/AxisAngleTest.java (hunk 4 above)
Comment (SylvainBertrand) [reply src/test/java/us/ihmc/euclid/axisAngle/AxisAngleTest.java:236]: I modified `EuclidCoreRandomTools`, the way orientations were generated weren't proper random. After fixing that, it broke a bunch of tests. Some tests aren't really affected besides changing the sequence of number being spit out by `Random.next...()` and some other are "explored" more because now orientations are "more" random. In turn, that made me crank up some tolerances. Some tests should prob be revisited, but I think for now it's fine

File: src/test/java/us/ihmc/euclid/matrix/RotationMatrixTest.java (hunk 5 above)
Comment (SylvainBertrand) [reply src/test/java/us/ihmc/euclid/matrix/RotationMatrixTest.java:1104]: Affected directly by the way orientations are now generated: they're "more" random than before. So I'm not surprised that in turn it reduces this number of iterations which represents how fast orientations "denormalize" with operations.
